frontend/packages/dev-console/src/components/pipelines/modals/common/PipelineResourceDropdown.tsx

```diff
@@ -0,0 +1,93 @@
+import * as React from 'react';
+import * as _ from 'lodash';
+import { FormikValues, useField, useFormikContext } from 'formik';
+import { Select, SelectOption } from '@patternfly/react-core';
+import { referenceForModel } from '@console/internal/module/k8s';
+import {
+  useK8sWatchResource,
+  WatchK8sResource,
+} from '@console/internal/components/utils/k8s-watch-hook';
+import { LoadingInline } from '@console/internal/components/utils';
+import { PipelineResourceModel } from '../../../../models';
+import { PipelineResourceKind } from '../../../../utils/pipeline-augment';
+import { CREATE_PIPELINE_RESOURCE } from './const';
+
+import './PipelineResourceDropdown.scss';
+
+type PipelineResourceDropdownProps = {
+  autoSelect?: boolean;
+  filterType: string;
+  name: string;
+  namespace: string;
+  selectedKey: string;
+};
+
+const PipelineResourceDropdown: React.FC<PipelineResourceDropdownProps> = (props) => {
+  const { autoSelect, filterType, name, namespace, selectedKey } = props;
+
+  const [isExpanded, setExpanded] = React.useState(false);
+  const { setFieldValue, setFieldTouched } = useFormikContext<FormikValues>();
+  const [, { touched }] = useField(name);
+
+  const resourceDefinition: WatchK8sResource = React.useMemo(
+    () => ({
+      isList: true,
+      namespace,
+      kind: referenceForModel(PipelineResourceModel),
+    }),
+    [namespace],
+  );
+  const [resources, loaded, error] = useK8sWatchResource<PipelineResourceKind[]>(
+    resourceDefinition,
+  );
+
+  const availableResources: PipelineResourceKind[] = resources.filter(
+    (resource) => resource.spec.type === filterType,
+  );
+
+  const canAutoSelect = autoSelect && !touched && loaded && !error;
+  React.useEffect(() => {
+    if (canAutoSelect) {
+      if (availableResources.length === 0) {
+        setFieldValue(name, CREATE_PIPELINE_RESOURCE);
+      } else {
+        setFieldValue(name, availableResources[0].metadata.name);
+      }
+      setFieldTouched(name);
+    }
+  }, [canAutoSelect, name, availableResources, setFieldTouched, setFieldValue]);
+
+  const options = [
+    { label: 'Create Pipeline Resource', value: CREATE_PIPELINE_RESOURCE },
+    ...availableResources.map((resource) => {
+      const resourceName = resource.metadata.name;
+      const url = _.get(_.find(resource.spec.params, ['name', 'url']), 'value', '');
```

Comment:
Why not use optional chaining here instead?

Reply:
Copy paste; didn't write the logic when I was refactoring. Will look to see about optional channing.